platforms/chibios/drivers/vendor/RP/RP2040/ps2_vendor.c

```diff
@@ -0,0 +1,335 @@
+        osalSysLockFromISR();
+        osalThreadResumeI(&tx_thread, MSG_OK);
+        osalSysUnlockFromISR();
+    }
+}
+
+void ps2_host_init(void) {
+    uint pio_idx = pio_get_index(pio);
+
+    hal_lld_peripheral_unreset(pio_idx == 0 ? RESETS_ALLREG_PIO0 : RESETS_ALLREG_PIO1);
+
+    state_machine = pio_claim_unused_sm(pio, true);
+    if (state_machine < 0) {
+        dprintln("ERROR: Failed to acquire state machine for PS/2!");
+        ps2_error = PS2_ERR_NODATA;
+        return;
+    }
+
+    uint offset = pio_add_program(pio, &ps2_program);
+
+    pio_sm_config c = pio_get_default_sm_config();
+    sm_config_set_wrap(&c, offset + PS2_WRAP_TARGET, offset + PS2_WRAP);
+    sm_config_set_sideset(&c, 1, false, false);
+
+    pio_sm_set_consecutive_pindirs(pio, state_machine, PS2_DATA_PIN, 2, false);
+    sm_config_set_clkdiv(&c, (float)clock_get_hz(clk_sys) / (200.0f * KHZ));
+    sm_config_set_set_pins(&c, PS2_DATA_PIN, 2);
+    sm_config_set_out_pins(&c, PS2_DATA_PIN, 1);
+    sm_config_set_out_shift(&c, true, false, 10);
+    sm_config_set_in_shift(&c, true, true, 11);
+    sm_config_set_jmp_pin(&c, PS2_CLOCK_PIN);
+    sm_config_set_in_pins(&c, PS2_DATA_PIN);
+
+    // clang-format off
+    iomode_t pin_mode = PAL_RP_PAD_IE |
+                        PAL_RP_GPIO_OE |
+                        PAL_RP_PAD_SLEWFAST |
+                        PAL_RP_PAD_DRIVE12 |
+                        PAL_RP_IOCTRL_OEOVER_DRVPERI |
+                        (pio_idx == 0 ? PAL_MODE_ALTERNATE_PIO0 : PAL_MODE_ALTERNATE_PIO1);
+    // clang-format on
+
+    palSetLineMode(PS2_DATA_PIN, pin_mode);
+    palSetLineMode(PS2_CLOCK_PIN, pin_mode);
+
+#ifdef BUFFERED_MODE_ENABLE
+    pio_set_irq0_source_enabled(pio, pis_sm0_rx_fifo_not_empty + state_machine, true);
+#endif
+
+    pio_sm_init(pio, state_machine, offset, &c);
+
+#if defined(PS2_PIO_USE_PIO1)
+    nvicEnableVector(RP_PIO1_IRQ_0_NUMBER, CORTEX_MAX_KERNEL_PRIORITY);
+#else
+    nvicEnableVector(RP_PIO0_IRQ_0_NUMBER, CORTEX_MAX_KERNEL_PRIORITY);
+#endif
+
+    pio_sm_set_enabled(pio, state_machine, true);
+}
+
+static int bit_parity(int x) {
+    x ^= x >> 16;
+    x ^= x >> 8;
+    x ^= x >> 4;
+    x ^= x >> 2;
+    x ^= x >> 1;
+    return (~x) & 1;
+}
+
+uint8_t ps2_host_send(uint8_t data) {
+    uint32_t frame = 0b1000000000;
+    frame          = frame | data;
+
+    if (bit_parity(data)) {
+        frame = frame | (1 << 8);
+    }
+
+    pio_sm_put(pio, state_machine, frame);
+
+    msg_t msg = MSG_OK;
+    osalSysLock();
+    if (pio_sm_is_tx_fifo_full(pio, state_machine)) {
+        pio_set_irq0_source_enabled(pio, pis_sm0_tx_fifo_not_full + state_machine, true);
+        msg = osalThreadSuspendTimeoutS(&tx_thread, TIME_MS2I(100));
+        if (msg < MSG_OK) {
+            pio_set_irq0_source_enabled(pio, pis_sm0_tx_fifo_not_full + state_machine, false);
+            ps2_error = PS2_ERR_NODATA;
+            return 0;
+        }
+    }
```

Comment:
Shouldn't that be a while loop that checks the tx fifo status again? Also if you early return in case of the timeout you have to unlock explicitly.

Reply:
About the unlock, nice catch, thanks! Will fix it asap.

About the tx fifo status, if it takes more than 100 ms to send the data to FIFO, something very bad must be going on in PIO (or anywhere else), thus the while loop infinitely (if the FIFO would stay still full), so I thought that it would be better to return nothing and try to tell to PS/2 driver that something went wrong (so, NODATA).

But yeah, if PS/2 "API" expects to those function to be blocking until there is not any data, then having it in while would be best. But for example, in https://github.com/qmk/qmk_firmware/blob/develop/drivers/ps2/ps2_interrupt.c#L161 , it basically timeouts after 25 tries. So, do you think it will be better to those functions to be always blocking, or they should time out? Thanks

Reply:
I was a bit to brief here. It should be non-blocking as you implemented it and also handle error properly as we don't want to block the main thread. What I was referring too was the pattern that I used in the [`sync_rx`/`sync_tx` functions]( https://github.com/qmk/qmk_firmware/blob/develop/platforms/chibios/drivers/vendor/RP/RP2040/serial_vendor.c#L200) where we break out of the while loop on timeout and exit the while loop after checking that we really reached the wanted fifo condition. But it is more a question of style not correctness.